### ISSUE TITLE
src/unix/threadpsx.cpp|wxThreadModule::OnExit thinks wxThread::Delete deletes something

### ISSUE THREAD

**Issue by fwr-ml** [wxWidgets/wxWidgets#23755]
To get that out of the way, 3.2.0, building static monolithic on Linux for Linux (gtk). After a quick look at the code in master, I'd say nothing's changed since.

In wxThreadModule::OnExit the following wrong assumptions about invariants seem to have been made:
- gs_allThreads would contain wxThread pointers to all and only to threads for which the decision for deletion had not been made, yet
- gs_nThreadsBeingDeleted would be the count of wxThreads for which the decision for self deletion has already been made(, but not yet executed)
- wxThread::Delete would have full effect on gs_allThreads and - where appropriate - gs_nThreadsBeingDeleted immediately

And, wxThreadModule::OnExit does its things in the wrong order. Should be:
- iterate over gs_allThreads and make the decision for deletion for every item
- wait on gs_condAllDeleted
- destroy globals like gs_mutexAllThreads

I'd say, wxThread::Delete should actually ensure the above assumptions and wxThreadModule::OnExit should reorder its steps.

Please correct me, if I'm wrong.

**Comment by vadz** [comment]
Sorry, what exactly is the problem you're trying to solve? All this is legacy code (all new programs ought to be using `std::thread` in any case) and while there are certainly problems in it, I'm not that eager to change anything there unless it's really necessary.

**Comment by fwr-ml** [comment]
Where I get to decide, I do use std::thread.
But I also use wxJoystick, which uses wxThread derived wxJoystickThread.

I'm currently in the process of making sure my application properly cleans up after itself. After I straightened out my user code enough to not case a crash in the midst of shutting down, wxw got to telling me, there were still wxThreads around.
Which led me to find out that wxThreadModule::OnExit in its remnants deleting loop calls wxThread::Delete on the same wxThread pointer repeatedly.
One wxThread (via wxJoystick) actually was unaccounted for. The other though is just being "deleted" in shutdown user code, but that doesn't take, when it comes to wxThreadModule::OnExit.

There's a workaround for me, sleeping a second after ```delete joystick_obj;```, but that might hide other problems in my code, too.

If it's deemed legacy code, should I change the title to "wxJoystick still uses misbehaving legacy code"? (honest question)

**Comment by vadz** [comment]
This joystick thread is indeed a problem. It would be nice to switch to using `std::thread` for it (and the sound playback thread in `src/unix/sound.cpp`), but this can only be done in master, not 3.2. There we'd have to fix the bug.

Unfortunately it's still not really clear to me what exactly the bug is and I don't have time to look into this now. It would definitely help a lot to have a simple example reproducing it -- it would also help to confirm that it's fixed later by running it under TSAN -- so if you could please make one, it would be very welcome. Of course, if you can provide a PR with the proposed fix too, this would be great. But we still need a test case to check that the problem is fixed.

**Comment by fwr-ml** [comment]
For an example it should suffice to create two wxJoystick objects in the app's OnInit and to delete them in the app's OnExit.
iirc, the objects would need to be constructed with IDs of actual devices, otherwise no threads would be created. And maybe, it'd need to be IDs of different devices - maybe not.

Then one should be able to observe, that
1) wxThreadModule::OnExit sees both as dangling threads
2) wxThreadModule::OnExit calls wxThread::Delete twice on the same wxThread pointer


I'm hesitant to propose a patch myself, for I don't even understand how that wxThread subsystem manages to get compiled into something that works at runtime at all.
I did not look at the build system, but had made an addition to interface/wx/thread.h, not yet to include/wx/thread.h though. src/unix/joystick.cpp saw the addition. src/unix/threadpsx.cpp did not.
Thus wxJoystickThread inherits from a wxThread definition without data members, does not reserve extra space for that subobject, but the implementation in threadpsx.cpp treats those objects as if they had data members according to the include/... definition.
I'm no C++ expert, but I'd think that should get undefined behaviour all over your face.

Aside from that, I guess changing the given system so that it would behave correctly seems to be not trivial.

As it is, wxThread::Delete only tries to arrange for the actual thread to finilize itselve, the latter then would also delete the wxThread object. For that Delete calls SetCancelFlag on the internal thread and tries to make sure it's active (start, if not started yet; resume if paused).
But Delete has  no immediate sideeffects, that wxThreadModule::OnExit would see.
So Delete should also call ScheduleThreadForDeletion and remove the wxThread pointer from gs_allThreads.
Then of course, all accounts of removing from gs_allThreads should first check, if it's still in there. AND all accounts of calling ScheduleThreadForDeletion should first check if gs_nThreadsBeingDeleted has already got incremented on account of the thread in question.
As it is, there's not enough data to do the latter check.

Then in wxThreadModule::OnExit:
First, iterate over gs_allThreads back to front calling Delete on the contained wxThread pointers. This whole loop should be enclosed with a mutex lock. The lock would need to be a recursive one or passed to a new Delete that expects the lock's reference - the old Delete would just make a lock and call the new Delete with that (new Delete then should probably check the lock for actually being locked and for locking the correct mutex (if possible)).
Second, do the wait on gs_condAllDeleted, as is - maybe with timeout; on timeout debug-log and abort.
Third, delete global objects, as is.

I can't say, if other paths of thread deletion would have to be adjusted, too, to stay compatibile.

**Comment by vadz** [comment]
Just to answer the obvious part: `interface` headers are used for documentation and wrappers generation only, they are not used at all for building the C++ library itself, only the header in `include` are used for this.

For the rest, it's not immediately obvious what should be done here to me neither and I'm unfortunately not sure when would I be able to look at this, so I can only hope you or somebody else can see a fix.

**Comment by fwr-ml** [comment]
You're right, interface/wx/thread.h does not get included in joystick.cpp. Just tried to make a case - didn't work out. I must have had jumped to conclusions there.
